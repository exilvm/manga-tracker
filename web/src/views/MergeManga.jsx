import fetch from 'cross-fetch';
import React from 'react';
import Typography from '@material-ui/core/Typography';
import Container from '@material-ui/core/Container';
import {makeStyles} from "@material-ui/core/styles";
import Button from "@material-ui/core/Button";
import Paper from '@material-ui/core/Paper'
import Grid from '@material-ui/core/Grid'
import Search from "../components/Search";
import PartialManga from "../components/PartialManga";

const host = process.env.HOST || 'http://localhost:3000';

const useStyles = makeStyles((theme) => ({
  root: {
    minHeight: '400px',
    minWidth: '900px',
    padding: theme.spacing(2),
  },
  mangaView: {
    width: '48%',
  },
  searchInput: {
    width: "100%",
    marginLeft: '1em',
  },
  inputRoot: {
    width: "100%",
    color: "inherit",
  },
  mergeButton: {
    padding: theme.spacing(2),
    textAlign: 'center',
  },
  errorMessage: {
    marginTop: theme.spacing(2),
  }
}));


function MergeManga(props) {
  const {
    isAuthenticated,
    user,
  } = props;

  const classes = useStyles();

  const [manga1, setManga1] = React.useState({});
  const [manga2, setManga2] = React.useState({});
  const [result, setResult] = React.useState({});

  const getMangaData = (manga_id, cb) => {
    fetch(`${host}/api/manga/${manga_id}`)
      .then(res => res.json())
      .then(cb)
      .catch(err => {
        console.error(err);
        cb({});
      })
  }

  const mergeManga = () => {
    if (!manga1.manga_id || !manga2.manga_id || manga1.manga_id == manga2.manga_id) return;
    fetch(`${host}/api/manga/merge/?base=${manga1.manga_id}&to_merge=${manga2.manga_id}`,
      {credentials: 'include', method: 'post'})
      .then(res => {
        if (res.status != 200) {
          setResult({error: true, message: `${res.status} ${res.statusText}`});
          return;
        }
        return res.json();
      })
      .then(json => {
        if (!json) return;
        setResult({message: `Moved ${json.alias_count} aliase(s) and ${json.chapter_count} chapter(s)`});
        setManga2({});
      })
      .catch(err => setResult({error: true, message: err.message}));
  }

  const renderItem = (setManga) => (option, index, props) => {
    return (
      <li key={index} >
        <div{...props} onClick={(e) => {
          e.currentTarget.parentElement.blur();
          getMangaData(option.manga_id, setManga);
        }}>{option.title}</div>
      </li>
    );
  }

  return (
    <Container maxWidth='xl'>
      <Paper className={classes.root}>
        <Grid
          container
          justify='space-between'
        >
          <div className={classes.mangaView}>
            <Search
              inputClasses={{
                input: classes.searchInput,
                root: classes.inputRoot,
              }}
              closeOnClick
              popperProps={{
                placement: 'bottom-start'
              }}
              renderItem={renderItem(setManga1)}

            />
            <PartialManga mangaData={manga1} showId/>
          </div>
          <div className={classes.mangaView}>
            <Search
              inputClasses={{
                input: classes.searchInput,
                root: classes.inputRoot,
              }}
              closeOnClick
              popperProps={{
                placement: 'bottom-start'
              }}
              renderItem={renderItem(setManga2)}
            />
            <PartialManga mangaData={manga2} showId />
          </div>
        </Grid>
        <Grid className={classes.mergeButton}>
          {manga1.manga_id && manga2.manga_id && manga1.manga_id !== manga2.manga_id &&
          <Button variant='contained' color='primary' onClick={mergeManga}>Merge</Button>}
          <Typography
            color={result.error ? 'error' : 'initial'}
            className={classes.errorMessage}
          >
            {result.message || null}
          </Typography>
        </Grid>
      </Paper>
    </Container>
  )
}
export default MergeManga;