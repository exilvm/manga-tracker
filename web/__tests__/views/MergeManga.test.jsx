import React from 'react';
import renderer from 'react-test-renderer';
import MergeManga from '../../src/views/MergeManga';

describe('Merge manga page should render correctly', () => {
  it('should render correctly by default', () => {
    const tree = renderer
      .create(<MergeManga />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
