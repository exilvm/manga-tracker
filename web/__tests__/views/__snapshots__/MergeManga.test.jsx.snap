// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Merge manga page should render correctly should render correctly by default 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthXl"
>
  <div
    className="MuiPaper-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-space-between"
    >
      <div
        className="makeStyles-mangaView-2"
      >
        <div
          className="makeStyles-search-7"
          onKeyDown={[Function]}
        >
          <div
            className="MuiInputBase-root makeStyles-inputRoot-4 MuiInputBase-adornedEnd"
            onClick={[Function]}
          >
            <input
              aria-label="search"
              autoCapitalize="none"
              autoComplete="off"
              className="MuiInputBase-input makeStyles-searchInput-3 MuiInputBase-inputAdornedEnd"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Search…"
              spellCheck={false}
              type="text"
            />
            <button
              className="MuiButtonBase-root MuiIconButton-root"
              disabled={false}
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        className="makeStyles-mangaView-2"
      >
        <div
          className="makeStyles-search-7"
          onKeyDown={[Function]}
        >
          <div
            className="MuiInputBase-root makeStyles-inputRoot-4 MuiInputBase-adornedEnd"
            onClick={[Function]}
          >
            <input
              aria-label="search"
              autoCapitalize="none"
              autoComplete="off"
              className="MuiInputBase-input makeStyles-searchInput-3 MuiInputBase-inputAdornedEnd"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Search…"
              spellCheck={false}
              type="text"
            />
            <button
              className="MuiButtonBase-root MuiIconButton-root"
              disabled={false}
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root makeStyles-mergeButton-5"
    >
      <p
        className="MuiTypography-root makeStyles-errorMessage-6 MuiTypography-body1"
      />
    </div>
  </div>
</div>
`;
